messages/ruby/lib/cucumber/messages/message/deserialization.rb

```diff
@@ -0,0 +1,142 @@
+          #   Cucumber::Messages::PickleTag.from_h({ name: "foo", ast_node_id: "abc-def" })  # => #<Cucumber::Messages::PickleTag:0x00007efda138cdb8 @name="foo", @ast_node_id="abc-def">
+          #
+          # It is available on the base Cucumber::Messages::Message class as long as the
+          # keys match one of the DTO class available with the messages
+          #
+          #   Cucumber::Messages::Message.from_h({ pickle_tag: { name: "foo", astNodeId: "abc-def" } })  # => #<Cucumber::Messages::PickleTag:0x00007efda1451a00 @name="foo", @ast_node_id="abc-def">
+          #
+          # If more than one message is part of the hash, it will return an array of messages
+          #
+          #   Cucumber::Messages::Message.from_h({
+          #     pickleTag: {name: "foo", ast_node_id: "abc-def"},
+          #     duration: {seconds: 1, nanos: 42 }
+          #   })
+          #   => [
+          #     #<Cucumber::Messages::PickleTag:0x00007efda6a2faa8 @name="foo", @ast_node_id="abc-def">,
+          #     #<Cucumber::Messages::Duration:0x00007efda6a24360 @seconds=1, @nanos=42>
+          #   ]
+          #
+          # It is recursive so embedded messages are also processed.
+          #
+          #   messages = { comment: { location: { line: 2 }, text: "comment" } }
+          #   Cucumber::Messages::Message.from_h(messages)  # => #<Cucumber::Messages::Comment:0x00007efda6abf888 @location=#<Cucumber::Messages::Location:0x00007efda6abf978 @line=2, @column=nil>, @text="comment">
+          #
+          #   gherkin_document = { uri: 'file:///...', comments: [{text: 'text comment'}, {text: 'another comment'}]}
+          #   Cucumber::Messages::GherkinDocument.from_h(gherkin_document)  # => #<Cucumber::Messages::GherkinDocument:0x00007efda11e6a90 ... @comments=[#<Cucumber::Messages::Comment:0x00007efda11e6e50 ..., #<Cucumber::Messages::Comment:0x00007efda11e6b58 ...>]>
+          #
+
+          def from_h(hash)
+            unless self.to_s === Cucumber::Messages::Message.to_s
+              return self.new(prepare_arguments(hash))
+            end
+
+            messages = hash.map do |key, value|
+              dto_name = camelize(key.to_s.dup)
+              dto_name[0] = key[0].upcase
+
+              dto = find_dto(dto_name)
+              arguments = prepare_arguments(value, dto: dto)
+
+              dto.new(arguments)
+            end
+
+            return messages.first if messages.length === 1
+
+            messages
+          end
+
+          private
+
+          def prepare_arguments(raw_arguments, dto: self)
+            raw_arguments.each do |key, value|
+              raw_arguments[key] = prepare_argument(key, value, dto: dto)
+            end
+
+            raw_arguments.transform_keys { |key| underscore(key.to_s).to_sym }
+          end
+
+          def find_dto(name)
+            Object.const_get("::Cucumber::Messages::#{name}")
```

Comment:
This won't work. You cannot assume that the property name has a corresponding message type. For example, one of the messages has a property called `implementation` (of type `Product`), but here you're trying to access a non-existent constant called `Implementation`